kotlin-coroutines-end/app/src/test/java/com/example/android/kotlincoroutines/CoroutinesTestRule.kt

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.android.kotlincoroutines
+
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import kotlinx.coroutines.test.resetMain
+import kotlinx.coroutines.test.setMain
+import org.junit.rules.TestWatcher
+import org.junit.runner.Description
+import java.util.concurrent.Executors
+
+/**
+ * Sets the main coroutines dispatcher for unit testing.
+ *
+ * Uses a TestCoroutineDispatcher if provided. Otherwise it uses a new single thread
+ * executor.
+ *
+ * See https://medium.com/androiddevelopers/easy-coroutines-in-android-viewmodelscope-25bffb605471
+ * and https://github.com/Kotlin/kotlinx.coroutines/tree/master/kotlinx-coroutines-test
+ */
+@ExperimentalCoroutinesApi
+class CoroutinesMainDispatcherRule(
+        private val testDispatcher: TestCoroutineDispatcher? = null
+) : TestWatcher() {
```

Comment:
Since this will be copy pasted, expose the TestCoroutineScope interface on this rule via property delegation, so tests can do

rule.runBlockingTest() 

without introducing a Dispatcher and Scope variable in every test class.

Reply:
nit; CoroutinesTestRule is a bit shorter to type and more accurate once you expose scope